### DISCUSSION TITLE
ZF1-Future in OpenMage

### DISCUSSION THREAD

**Discussion by addison74** [OpenMage/magento-lts#2194] [General]
I know that there have been discussions on this topic, so I intend to complete this important project to replace ZF1 in OpenMage with ZF1-Future which is an active project and where new versions are periodically released based on a dedicated community. It was a pleasant surprise to me they have few issues, the reported ones are solved quickly, valid also for PRs. In addition it is compatible with PHP 8.1.

I recently found that parts of the ZF1 code in OpenMage no longer match current configurations, for example the TLS protocol used in SMTP is still limited to version 1.0, although for several years most MTA's no longer accept connections under version 1.2 for security reasons.

I had the opportunity these days to analyze the implications of this change and my opinion is that we could need a dedicated repository as part of OpenMage (for example openmage: zend) where a custom version of ZF1-Future will be provided which will be added to any new release of OpenMage and that we will maintain entirely.

At the first stage we will have to analyze the changes in the two directories used by this framework and which can be found here:

**/app/code/core/Zend
/lib/Zend**

We are not going to copy them simply because as you will see in the following posts there are differences that need to be analyzed in detail. intensive testing over a longer period is absolutely necessary. The files maybe need to be adapted and then with each new ZF1-Future version to operate the changes in the repository, most likely they will be some directly by copying and others will require manual changes.

**Comment by addison74** [comment]
This a visual comparison for **/lib/Zend** directory between OpenMage 20.0.14 and ZF1-Future 1.21.2. The report can be downloaded from here: 

https://mega.nz/file/2UdHzZpB#luas9xRvYGENfMysJgH2TeLH2ZAyTxFSG6BkiYt4K08

This is the list of files that are on one directory and not the other (they have no correspondent).

**Files and directories found in OpenMage but not in ZF1-Future**
/Gdata/Analytics/Extension/Goal.php
/Gdata/YouTube
/Gdata/YouTube/Extension
/Gdata/YouTube.php
/Tool/Framework/Provider/DocblockManifestable.php
/replace_recursive.php

**Files and directories found in Zf1-Future but not in OpenMage**
/Db/Adapter/Pdo/Sqlsrv.php
/Db/Schema
/Db/Schema/AbstractChange.php
/Db/Schema/Exception.php
/Db/Schema/Manager.php
/Gdata/Analytics/Goal.php
/Tool/Framework/Provider/DocblockManifestInterface.php
/Tool/DatabaseSchemaProvider.php

**Comment by addison74** [comment]
This a visual comparison for **/app/code/core/Zend** directory between OpenMage 20.0.14 and ZF1-Future 1.21.2. The report can be downloaded from here: 

https://mega.nz/file/rEEzQKjb#lnZUK5DZI6m11gkRl91oPa9TCGioarGP411GTtv3fYI

I ask myself the natural question if we opt for a repository that we maintain based on ZF1-Future do we still need this directory /app/code/core/Zend since we can operate all the changes in /lib/Zend? 

Some of you have even proposed drafts for cleaning up /app directory and leave all files in /lib/Zend. I guess the Magento team was intended to keep the original ZF1 in /lib/Zend for easy updates and create a custom version in /app/code/core/Zend. Based on Magento fallback the files were served from both directories.

**Comment by addison74** [comment]
**COMPARISON for /app/code/core/Zend/Controller/Request/Http.php**

**Open Mage - line 1037**
```php
public function getScheme()
{
    return return (strtolower((string)$this->getServer('HTTPS')) == 'on') || $this->getServer('HTTP_X_FORWARDED_PROTO') == 'https' ?
    self::SCHEME_HTTPS :
    self::SCHEME_HTTP;
}
```

**ZF1-Future -  line 1037**

```php
public function getScheme()
{
    return ($this->getServer('HTTPS') == 'on') ? self::SCHEME_HTTPS : self::SCHEME_HTTP;
}
```

In this PR #1462 we used both headers $_SERVER['HTTP_X_FORWARDED_PROTO'] and $_SERVER['HTTPS']. 

I think we need to make a connection between these two headers somewhere in the OM code so that will not be an issue if the ZF1-Future version is used from now on and which does not include Forwarded-Proto like in OM. It can be done in .htaccess too and use only HTTPS = on if the connection is secured. And we should have a line in README before implementing ZF1-Future.

**Comment by addison74** [comment]
These files have comments inside from the Magento team:

**/app/code/core/Zend/Db/Select.php**
**/app/code/core/Zend/Db/Statement.php**
There are a few small changes the Magento team did. Only with ZF1-Future files there are errors in different sections of the Backend. The issue could be created by function _stripQuoted($sql).

**/app/code/core/Zend/Filter/PregReplace.php**

```php
/**
 * This class replaces default Zend_Filter_PregReplace because of problem described in MPERF-10057
 * The only difference between current class and original one is overwritten implementation of filter method and add new
 * method _isValidMatchPattern
 *
```

**/app/code/core/Zend/Form/Decorator/Form.php**

```php
/**
 * This class replaces default Zend_Form_Decorator_Form because of problem described in MPERF-9707/MPERF-9769
 * The only difference between current class and original one is overwritten implementation of render method
 *
```

**/app/code/core/Zend/Locale/Math/PhpMath.php**
```php
/**
 * This class replaces default Zend_Locale_Math_PhpMath because of issues described in MPERF-10261 and MPERF-10262
 * The only difference between current class and original one is overwritten implementation of Sub method
 *
```

**/app/code/core/Zend/Serializer/Adapter/PhpCode.php**
```php
/**
 * This class replaces default Zend_Serializer_Adapter_PhpCode because of problem described in MPERF-9450
 * The only difference between current class and original one is overwritten implementation of unserialize method
 * 
```

**/app/code/core/Zend/Validate/EmailAddress.php**
```php
/**
 * This class replaces default Zend_Validate_Email_Address because of issues described in MPERF-9688 and MPERF-9689
 * The only difference between current class and original one is overwritten implementation of _validateLocalPart method
 * 

```

**/app/code/core/Zend/Validate/HostName.php**
This file in OM has at the end function checkDns Records which doesn't exist in ZF1-Future

**/app/code/core/Zend/Date.php**
```php
/**
 * This class replaces default Zend_Date because of problem described in Jira ticket MAGE-4872
 * The only differences between current class and original one is overwritten implementation of mktime method
 * 
```

It is obvious that the first evaluation for this major replacement must start from the comparison of the files in the /app/code/core/Zend directory with the one in ZF1-Future. The comparison provided by me is more than enough for an evaluation. Maybe we could have a customized version of the framework.

**Reply by justinbeaty** [reply]
The fixes could potentially be sent as a PR to upstream ZF1-Future as well, depending on if they'd accept them.

**Reply by addison74** [reply]
I agree. In OM we can update whenever a new version of ZF1-Future is released, copying the framework to /lib/Zend. At the same time in /app/code/core/Zend there will be those specific changes for OM to work properly. These files will need to be reviewed whenever a new version is released, but it is an easy task as long as there is a limited number of files (I identified 16 files) and we can immediately find out if the new ZF1-Future versions are aimed directly at them.

In order to take this platform as far as possible and keep up with the new PHP version, it needs ZF1-Future. I detailed in a post that TLS is left behind to version 1.0 for current OM. In the near future most browsers will require connections from TLSv1.2 onwards.

**Reply by sreichel** [reply]
> The fixes could potentially be sent as a PR to upstream ZF1-Future as well, depending on if they'd accept them.

This already happend for some changes. I just wait for response before i continue.

**Comment by sreichel** [comment]
@ADDISON74 PR for ZF1 is ready ... but check this tomorrow :+1:

**Reply by addison74** [reply]
"The Eagle has landed". 

Nice work as usual.

**Comment by addison74** [comment]
Thank you to all who made this possible. ZF1-Future was implemented into OpenMage.
